pipeline/trusted-resources/pkg/trustedtask/trustedpipelinerun.go

```diff
@@ -0,0 +1,188 @@
+	_ apis.Validatable = (*TrustedPipelineRun)(nil)
+)
+
+// Validate the PipelineRunRef referred pipeline and task are tampered or not.
+func (pr *TrustedPipelineRun) Validate(ctx context.Context) (errs *apis.FieldError) {
+	// TODO: validate only on create operation.
+
+	k8sclient := kubeclient.Get(ctx)
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return apis.ErrGeneric(err.Error())
+	}
+	tektonClient, err := versioned.NewForConfig(config)
+	if err != nil {
+		return apis.ErrGeneric(err.Error())
+	}
+
+	if errs := errs.Also(pr.verifyPipelineRun(ctx, k8sclient, tektonClient)); errs != nil {
+		return errs
+	}
+	return nil
+}
+
+// SetDefaults is not used.
+func (pr *TrustedPipelineRun) SetDefaults(ctx context.Context) {
+}
+
+func (pr *TrustedPipelineRun) verifyPipelineRun(
+	ctx context.Context,
+	k8sclient kubernetes.Interface,
+	tektonClient versioned.Interface,
+) (errs *apis.FieldError) {
+	logger := logging.FromContext(ctx)
+	logger.Info("Verifying PipelineRun")
+
+	serviceAccountName := os.Getenv("WEBHOOK_SERVICEACCOUNT_NAME")
+	if serviceAccountName == "" {
+		serviceAccountName = "tekton-verify-task-webhook"
+	}
+
+	// for inline pipeline we verify the tasks
+	if pr.Spec.PipelineRef == nil {
+		for _, t := range pr.Spec.PipelineSpec.Tasks {
+			if t.TaskRef == nil {
+				continue
+			}
+
+			fn, err := taskrun.GetTaskFunc(ctx, k8sclient, tektonClient, t.TaskRef, pr.Namespace, serviceAccountName)
+			if err != nil {
+				return apis.ErrGeneric(err.Error(), "spec", "taskRef")
+			}
+
+			if err := verifyTask(ctx, t.TaskRef.Name, k8sclient, fn); err != nil {
+				return apis.ErrGeneric(err.Error(), "spec", "taskRef")
+			}
+		}
+		return nil
+	}
+
+	cfg := config.FromContextOrDefaults(ctx)
+	cfg.FeatureFlags.EnableTektonOCIBundles = true
+	ctx = config.ToContext(ctx, cfg)
+
+	fn, err := resources.GetPipelineFunc(ctx, k8sclient, tektonClient, &pr.PipelineRun)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "spec", "PipelineRef")
+	}
+	resolvedPipeline, err := fn(ctx, pr.Spec.PipelineRef.Name)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "spec", "PipelineRef")
+	}
+
+	pipeline, signature, err := preparePipeline(resolvedPipeline)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "spec", "PipelineRef")
+	}
+
+	v, err := verifier(ctx, pipeline.ObjectMeta.Annotations, k8sclient)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "PipelineRef")
+	}
+
+	if err := VerifyInterface(ctx, pipeline, v, signature); err != nil {
+		return apis.ErrGeneric(err.Error(), "PipelineRef")
+	}
+
+	for _, t := range pipeline.Spec.Tasks {
+		if t.TaskRef == nil {
+			continue
+		}
+
+		fn, err := taskrun.GetTaskFunc(ctx, k8sclient, tektonClient, t.TaskRef, pr.Namespace, serviceAccountName)
+		if err != nil {
+			return apis.ErrGeneric(err.Error(), "spec", "taskRef")
+		}
+
+		if err := verifyTask(ctx, t.TaskRef.Name, k8sclient, fn); err != nil {
+			return apis.ErrGeneric(err.Error(), "spec", "taskRef")
+		}
+
+	}
```

Comment:
This looks duplicated with the code above - would it be possible to pull this into a common func?

Reply:
yes verify task and pipeline share the same logic, but they are using different getfunc and have different data type, it is not so clear to me how to extract this pattern into a common one, 🤔

Reply:
got it! makes sense.